text/0000-third-party-web-components-support.md

```diff
@@ -0,0 +1,155 @@
+```
+
+### Passing data
+
+By convention, custom elements are assumed to be backed by LWC components and values are set through
+their properties. Given that a native web component might expect data passed as an attribute or as a
+property, we would need to support both mechanisms. There are multiple approaches we could take
+here, and the following options are how some popular frameworks do it today. We could default to
+either attributes or properties and introduce a directive that would act as a signal to the engine
+to use one or the other, we could implement a runtime heuristic to choose one over the other, or we
+could default to attributes and set properties when setting a data structure.
+
+#### Proposed solution
+
+When passing data to a native web component, we can take a heuristical approach where if the
+property exists, the property is set. Otherwise the attribute is set. While it may be advantageous
+to require an explicit signal in the form of a directive (both in terms of static analysis and
+performance), the proposed solution strikes a good balance between developer experience and API
+semantics. This heuristical approach will require a check on the element's prototype to see whether
+a property is defined. This will incur a runtime performance cost, which can be amortized over the
+lifetime of the component by caching the result of the check.
+
+### Events
+
+Today, LWC's declarative event bindings only support lowercase events. Events using anything other
+than lowercase can be listened for using the imperative `addEventListener()` API. There are several
+historical discussions (e.g., [#1811][1811] and [#1904][1904]) about relaxing this constraint;
+however, the current behavior would not need to change in order to support native web components. It
+should be noted that adding support for events has always been about improved interoperability, as
+the browser's imperative API has no restriction regarding the event type.
+
+EventTarget polyfilled methods and properties such as `add/removeEventListener` and `target` go
+through the synthetic shadow polyfill due to those methods and properties being globally patched.
+This is the current status quo for native shadow roots and should not be an issue for supporting
+third party web components.
+
+### Serverside rendering (SSR)
+
+If an external element is marked with a directive that identifies it as such, its custom element
+should be rendered and its attributes should be set. Properties will not be set and no attempt will
+be made to render the subtree.
+
+### Custom element conflicts
+
+In environments like the Salesforce platform which employ a multitenant architecture, developers
+will certainly run into custom element conflicts when given the freedom to register their own
+components. This can be managed in some environments (i.e., internal developers or applications
+consuming OSS LWC) but will require virtualization per sandbox before we can GA the feature.

```

Comment:
I don't think LWC should care here... this is a platform thing. Also, I don't think virtualization capabilities by locker should gate this because today 2nd party and 3rd party components cannot register componnets anyway, so it should be fine... the only problem is really a platform registered component that can now be created by a namespace, which is fine I believe, we don't expect a bunch of people to register components in system mode anyways, and they can get to them by flashlight if that the case... I don't know.